modules/images/image-loading-optimization/class-ilo-url-metrics-group-collection.php

```diff
@@ -0,0 +1,241 @@
+<?php
+/**
+ * Image Loading Optimization: ILO_URL_Metrics_Group_Collection class
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Collection of URL groups according to the breakpoints.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+final class ILO_URL_Metrics_Group_Collection {
+
+	/**
+	 * URL metrics groups.
+	 *
+	 * The number of groups corresponds to one greater than the number of
+	 * breakpoints. This is because breakpoints are the dividing line between
+	 * the groups of URL metrics with specific viewport widths. This extends
+	 * even to when there are zero breakpoints: there will still be one group
+	 * in this case, in which every single URL metric is added.
+	 *
+	 * @var ILO_URL_Metrics_Group[]
+	 * @phpstan-var non-empty-array<ILO_URL_Metrics_Group>
+	 */
+	private $groups;
+
+	/**
+	 * Breakpoints in max widths.
+	 *
+	 * @var int[]
+	 * @phpstan-var int<1, max>[]
+	 */
+	private $breakpoints;
+
+	/**
+	 * Sample size for URL metrics for a given breakpoint.
+	 *
+	 * @var int
+	 * @phpstan-var int<1, max>
+	 */
+	private $sample_size;
+
+	/**
+	 * Freshness age (TTL) for a given URL metric.
+	 *
+	 * @var int
+	 * @phpstan-var int<0, max>
+	 */
+	private $freshness_ttl;
+
+	/**
+	 * Constructor.
+	 *
+	 * @throws InvalidArgumentException When an invalid argument is supplied.
+	 *
+	 * @param ILO_URL_Metric[] $url_metrics   URL metrics.
+	 * @param int[]            $breakpoints   Breakpoints in max widths.
+	 * @param int              $sample_size   Sample size for the maximum number of viewports in a group between breakpoints.
+	 * @param int              $freshness_ttl Freshness age (TTL) for a given URL metric.
+	 */
+	public function __construct( array $url_metrics, array $breakpoints, int $sample_size, int $freshness_ttl ) {
+		// Set breakpoints.
+		sort( $breakpoints );
+		$breakpoints = array_values( array_unique( $breakpoints, SORT_NUMERIC ) );
+		foreach ( $breakpoints as $breakpoint ) {
+			if ( $breakpoint <= 0 ) {
```

Comment:
See above, how is the maximum / infinite value handled? The last breakpoint group would not really have a maximum. Does that value need to be provided or is it handled automatically? I just wonder whether additional validation is needed.

Reply:
I guess I was thinking for the last group you wouldn't add the +1. That would mean every breakpoint represents a group, while right now there's always one less breakpoint than there are groups.

I guess both makes sense from different perspectives, so the current approach WFM.